ql/src/java/org/apache/hadoop/hive/ql/ppd/OpProcFactory.java

```diff
@@ -782,6 +790,89 @@ protected ExprWalkerInfo mergeChildrenPred(Node nd, OpWalkerInfo owi,
+    List<ExprNodeDesc> deterministicPreds = new ArrayList<ExprNodeDesc>();
+    Iterator<List<ExprNodeDesc>> iterator1 = pushDownPreds.values().iterator();
+    while (iterator1.hasNext()) {
+      for (ExprNodeDesc pred : iterator1.next()) {
+        deterministicPreds = ExprNodeDescUtils.split(pred, deterministicPreds);
+      }
+    }
+
+    if (deterministicPreds.isEmpty()) {
+      return null;
+    }
+
+    List<ExprNodeDesc> nondeterministicPreds = new ArrayList<ExprNodeDesc>();
+    Iterator<List<ExprNodeDesc>> iterator2 = unPushDownPreds.values().iterator();
+    while (iterator2.hasNext()) {
+      for (ExprNodeDesc pred : iterator2.next()) {
+        nondeterministicPreds = ExprNodeDescUtils.split(pred, nondeterministicPreds);
+      }
+    }
+
+    assert !nondeterministicPreds.isEmpty();
+
+    ExprNodeDesc deterministicCondn = ExprNodeDescUtils.mergePredicates(deterministicPreds);
+    ExprNodeDesc nondeterministicCondn = ExprNodeDescUtils.mergePredicates(nondeterministicPreds);
+
+    Operator<FilterDesc> deterministicFilter =
+        OperatorFactory.get(new FilterDesc(deterministicCondn, false), new RowSchema(inputRS.getSignature()));
+
+    deterministicFilter.setChildOperators(new ArrayList<Operator<? extends OperatorDesc>>());
+    deterministicFilter.getChildOperators().add(op);
+
+    List<Operator<? extends OperatorDesc>> originalParents = op
+        .getParentOperators();
+    for (Operator<? extends OperatorDesc> parent : originalParents) {
+      List<Operator<? extends OperatorDesc>> childOperators = parent
+          .getChildOperators();
+      int pos = childOperators.indexOf(op);
+      childOperators.remove(pos);
+      childOperators.add(pos, deterministicFilter);
+
+      int pPos = op.getParentOperators().indexOf(parent);
+      deterministicFilter.getParentOperators().add(pPos, parent);
+    }
+
+    op.getParentOperators().clear();
+    op.getParentOperators().add(deterministicFilter);
+    op.getConf().setPredicate(nondeterministicCondn);
+
+    if (HiveConf.getBoolVar(owi.getParseContext().getConf(),
+        HiveConf.ConfVars.HIVEPPDREMOVEDUPLICATEFILTERS)) {
+      // remove the candidate filter ops
+      for (FilterOperator fop : owi.getCandidateFilterOps()) {
+        List<Operator<? extends OperatorDesc>> children = fop.getChildOperators();
+        List<Operator<? extends OperatorDesc>> parents = fop.getParentOperators();
+        for (Operator<? extends OperatorDesc> parent : parents) {
+          parent.getChildOperators().addAll(children);
+          parent.removeChild(fop);
+        }
+        for (Operator<? extends OperatorDesc> child : children) {
+          child.getParentOperators().addAll(parents);
+          child.removeParent(fop);
+        }
+      }
+      owi.getCandidateFilterOps().clear();
+    }
+
+    ewi = ExprWalkerProcFactory.extractPushdownPreds(owi, op,
+        deterministicFilter.getConf().getPredicate());
+
+    owi.putPrunedPreds(deterministicFilter, ewi);
```

Comment:
the deterministicFilter should add to OpWalkerInfo.candidateFilterOps and no need to extractPushdownPreds again

Reply:
there is only to get ewi for push down latter

Reply:
Looking the codes, an temporary filter op is inserted and push the pruned predicates into current context. but if hive.ppd.remove.duplicatefilters is enabled, a redundant evaluation would happen on the pruned predicates, which composes the deterministicFilter.

Reply:
I think it is normal situation. for if hive.ppd.remove.duplicatefilters is enabled, we could produce other redundant filter which created by method createFilter. Therefore, I don't think it's worth worry

Reply:
the reason I use ExprWalkerProcFactory.extractPushdownPreds is that we need split candidates for situation may occur that there contains AND, also, we need assign deterministic predicates to  pushdownPreds which in ExprWalkerInfo.

Reply:
The and case has been processed when extractPushdownPreds,  ExprWalkerProcFactory.extractPushdownPreds -> ExprWalkerProcFactory.extractFinalCandidates. https://github.com/apache/hive/blob/6002c510113d9a6aa87159c7386f2a8a4747405b/ql/src/java/org/apache/hadoop/hive/ql/ppd/ExprWalkerProcFactory.java#L308-L328

Reply:
yes, but there is a method ExprNodeDescUtils.mergePredicates was invoked above, therefore,there may contains AND.